File: src/Infrastructure/Outbox.php
<?php
declare(strict_types=1);
namespace App\Infrastructure;
use App\Observability\OperationsService;
final class Outbox
{
    public function __construct(private Database $db, private ?\App\Settings\Vault $vault=null) {}
    /** Higher priority = drained first. Personal Telegram replies/answers outrank mass broadcasts. */
    private const PRIORITY = [
        'payment.verify' => 100,
        'payment.event.process' => 100,
        'payment.create' => 90,
        'topup.create' => 90,
        'topup.after' => 90,
        'referral.topup' => 90,
        'subscription.provision' => 80,
        'subscription.extend' => 80,
        'subscription.renew' => 80,
        'subscription.traffic' => 80,
        'subscription.devices' => 80,
        'gift.create' => 70,
        'telegram.send' => 60,
        'telegram.answer' => 60,
        'compensation.run' => 50,
        'compensation.grant' => 50,
        'broadcast.send' => 10,
        'broadcast.run' => 5,
    ];

    public function priority(string $topic): int
    {
        return self::PRIORITY[$topic] ?? 30;
    }

    public function enqueue(string $topic, string $key, array $payload, int $delay = 0): void
    {
        $json=json_encode($payload, JSON_THROW_ON_ERROR);
        if($this->vault)$json='enc:'.$this->vault->seal('outbox:'.$topic,$json);
        $this->db->execute('INSERT INTO outbox(id,topic,dedup_key,payload,priority,available_at,created_at) VALUES(?,?,?,?,?,?,?) ON CONFLICT(dedup_key) DO NOTHING', [Database::id(),$topic,$key,$json,$this->priority($topic),time()+$delay,time()]);
    }
    public function runOne(callable $handler): bool
    {
        $job = $this->db->transaction(function () {
            $lock = $this->db->postgres() ? ' FOR UPDATE SKIP LOCKED' : '';
            $job = $this->db->one("SELECT * FROM outbox WHERE (status='pending' AND available_at<=?) OR (status='processing' AND locked_until<=?) ORDER BY priority DESC, created_at,id LIMIT 1".$lock, [time(),time()]);
            if (!$job) return null;
            $job['lock_token'] = Database::id();
            $this->db->execute("UPDATE outbox SET status='processing', attempts=attempts+1, locked_until=?, lock_token=? WHERE id=?", [time()+120,$job['lock_token'],$job['id']]);
            return $job;
        });
        if (!$job) return false;
        $payload=[];
        $operations=new OperationsService($this->db);
        $operation=$job['correlation_id'] ? $this->db->one('SELECT id FROM operations WHERE correlation_id=?',[$job['correlation_id']]) : null;
        if($operation)$operations->event($operation['id'],'outbox.started','processing','Outbox job started',['metadata'=>['topic'=>$job['topic'],'attempt'=>(int)$job['attempts']+1]]);
        try {
            $json=$job['payload'];
            if(str_starts_with($json,'enc:'))$json=$this->vault?->open('outbox:'.$job['topic'],substr($json,4))??throw new \RuntimeException('Outbox decryption unavailable');
            $payload=json_decode($json,true,512,JSON_THROW_ON_ERROR);
            $handler($job['topic'], $payload);
            $this->db->execute("UPDATE outbox SET status='done',locked_until=NULL,last_error=NULL,payload='{}' WHERE id=? AND lock_token=?", [$job['id'],$job['lock_token']]);
            if($operation)$operations->event($operation['id'],'outbox.completed','success','Outbox job completed',['metadata'=>['topic'=>$job['topic']]]);
        } catch (JobDeferred $e) {
            // Maintenance and global safety mode are expected operational
            // states, not failures. Do not exhaust attempts or dead-letter a
            // valid money/provisioning command while an integration is paused.
            $this->db->execute("UPDATE outbox SET status='pending',attempts=CASE WHEN attempts>0 THEN attempts-1 ELSE 0 END,available_at=?,locked_until=NULL,lock_token=NULL,last_error=NULL WHERE id=? AND lock_token=?",[
                time()+max(1,$e->delaySeconds),$job['id'],$job['lock_token']
            ]);
            if($operation)$operations->event($operation['id'],'outbox.deferred','processing','Outbox job deferred by operational control',['metadata'=>['topic'=>$job['topic']]]);
        } catch (JobPermanentFailure $e) {
            // Terminal, non-recoverable failure (e.g. Telegram permanently
            // refuses a chat). Dead-letter now without spending retry budget so
            // a mass broadcast of undeliverable recipients never clogs the
            // queue or stalls the worker. This is an expected outcome (blocked
            // chat), not an anomaly worth a job.failed alert.
            $this->db->execute("UPDATE outbox SET status='dead',available_at=?,locked_until=NULL,last_error=? WHERE id=? AND lock_token=?", [time(), 'permanent:'.get_class($e), $job['id'], $job['lock_token']]);
            if ($operation) $operations->event($operation['id'], 'outbox.dead', 'warning', 'Outbox job permanently failed', ['metadata' => ['topic' => $job['topic'], 'error_class' => get_class($e)]]);
        } catch (\Throwable $e) {
            $attempt = (int)$job['attempts']+1;
            // Never persist raw HTTP errors: they may contain tokens or subscription URLs.
            $this->db->execute('UPDATE outbox SET status=?, available_at=?, locked_until=NULL,last_error=? WHERE id=? AND lock_token=?', [$attempt>=8?'dead':'pending',time()+min(3600,2**$attempt)+random_int(0,5),get_class($e),$job['id'],$job['lock_token']]);
            if (in_array($job['topic'],['subscription.provision','subscription.extend'],true) && isset($payload['subscription_id'])) {
                $state=$attempt>=8?'failed':'retry';
                $this->db->execute('UPDATE provisioning_accounts SET state=?,last_error=?,updated_at=? WHERE subscription_id=? AND state<>\'active\'',[
                    $state,get_class($e),time(),(string)$payload['subscription_id']
                ]);
            }
            if($operation)$operations->event($operation['id'],'outbox.retry','warning','Outbox retry scheduled',['metadata'=>['topic'=>$job['topic'],'attempt'=>$attempt,'error_class'=>get_class($e)]]);
            error_log(json_encode(['event'=>'job.failed','job_id'=>$job['id'],'type'=>get_class($e),'attempt'=>$attempt]));
        }
        return true;
    }
}

File: ruby_app/zeleboba/infrastructure/outbox.rb
# frozen_string_literal: true

require "json"

require_relative "database"
require_relative "job_error"

module Zeleboba
  module Infrastructure
    class Outbox
      PRIORITY = {
        "payment.create" => 90,
        "subscription.provision" => 80,
        "telegram.send" => 60
      }.freeze

      def initialize(db)
        @db = db
      end

      def enqueue(topic, key, payload, delay: 0)
        raise ArgumentError, "Outbox topic is required" if topic.to_s.empty?
        raise ArgumentError, "Outbox deduplication key is required" if key.to_s.empty?

        @db.execute(
          "INSERT INTO outbox(id,topic,dedup_key,payload,priority,available_at,created_at) VALUES(?,?,?,?,?,?,?) ON CONFLICT(dedup_key) DO NOTHING",
          [Database.id, topic, key, JSON.generate(payload), priority(topic), Time.now.to_i + delay, Time.now.to_i]
        )
      end

      def run_one
        job = @db.transaction do
          reclaim_expired_leases!
          row = @db.one(
            "SELECT * FROM outbox WHERE status='pending' AND available_at<=? ORDER BY priority DESC, created_at,id LIMIT 1#{@db.lock}",
            [Time.now.to_i]
          )
          next nil unless row

          token = Database.id
          @db.execute(
            "UPDATE outbox SET status='processing', attempts=attempts+1, locked_until=?, lock_token=? WHERE id=? AND status='pending'",
            [Time.now.to_i + lease_seconds, token, row["id"]]
          )
          row.merge("lock_token" => token)
        end
        return false unless job

        yield(job["topic"], JSON.parse(job["payload"], symbolize_names: false))
        @db.execute("UPDATE outbox SET status='done',locked_until=NULL,last_error=NULL,payload='{}' WHERE id=? AND lock_token=?", [job["id"], job["lock_token"]])
        true
      rescue JobDeferred => e
        retry_job(job, e.message, delay: e.delay) if job
        false
      rescue JobPermanentFailure => e
        dead_letter(job, e.message) if job
        false
      rescue StandardError => e
        if job
          attempt = job["attempts"].to_i + 1
          attempt >= max_attempts ? dead_letter(job, "#{e.class}: #{e.message}") : retry_job(job, "#{e.class}: #{e.message}", delay: backoff(attempt))
        end
        raise
      end

      def priority(topic)
        PRIORITY.fetch(topic, 30)
      end

      private

      def reclaim_expired_leases!
        @db.execute(
          "UPDATE outbox SET status='pending',locked_until=NULL,lock_token=NULL,last_error=COALESCE(last_error,'worker lease expired') WHERE status='processing' AND locked_until<?",
          [Time.now.to_i]
        )
      end

      def retry_job(job, message, delay:)
        @db.execute(
          "UPDATE outbox SET status='pending',available_at=?,locked_until=NULL,lock_token=NULL,last_error=? WHERE id=? AND lock_token=?",
          [Time.now.to_i + delay, message.to_s[0, 500], job["id"], job["lock_token"]]
        )
      end

      def dead_letter(job, message)
        @db.execute(
          "UPDATE outbox SET status='dead',locked_until=NULL,lock_token=NULL,last_error=? WHERE id=? AND lock_token=?",
          [message.to_s[0, 500], job["id"], job["lock_token"]]
        )
      end

      def lease_seconds = 120
      def max_attempts = 8
      def backoff(attempt) = [3600, 2**attempt].min
    end
  end
end
